File: skills/agent-swarm/tests/manual_acp_handshake.py
#!/usr/bin/env python3
"""Probe a real ACP Agent through the official SDK without Runtime identity."""

import argparse
import asyncio
import json
import pathlib
import sys


SCRIPTS_DIR = pathlib.Path(__file__).resolve().parents[1] / "scripts"
if str(SCRIPTS_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPTS_DIR))

from backends.acp.client import AgentSwarmClient, connect_agent
from backends.acp.dependencies import activate as activate_bundled_sdk
from backends.acp.permissions import PermissionDecision
from backends.acp.processes import process_group_alive, terminate_process_group


def _dump(value):
    if value is None:
        return None
    return value.model_dump(mode="json", by_alias=True, exclude_none=True)


async def probe(args):
    activate_bundled_sdk()
    from acp import PROTOCOL_VERSION, RequestError, text_block
    from acp.schema import ClientCapabilities, Implementation

    updates = []
    process = None
    connection = None
    report = {"ok": False}

    def session_update(session_id, update):
        updates.append(
            {
                "sessionId": session_id,
                "type": getattr(update, "session_update", type(update).__name__),
            }
        )

    try:
        process = await asyncio.create_subprocess_exec(
            args.command,
            *args.command_arg,
            cwd=args.cwd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            start_new_session=True,
        )
        if process.stdin is None or process.stdout is None:
            raise RuntimeError("ACP Agent did not expose stdio streams")
        connection = connect_agent(
            AgentSwarmClient(
                permission_handler=lambda request: PermissionDecision(None, False),
                session_update_handler=session_update,
            ),
            process.stdin,
            process.stdout,
        )
        initialized = await asyncio.wait_for(
            connection.initialize(
                protocol_version=PROTOCOL_VERSION,
                client_capabilities=ClientCapabilities(),
                client_info=Implementation(
                    name="agent-swarm-handshake", version="1"
                ),
            ),
            timeout=30,
        )
        report["initialized"] = {
            "protocolVersion": initialized.protocol_version,
            "agentInfo": _dump(initialized.agent_info),
            "agentCapabilities": _dump(initialized.agent_capabilities),
            "authMethods": [
                {key: getattr(method, key, None) for key in ("id", "name", "type")}
                for method in initialized.auth_methods or []
            ],
        }
        session = await asyncio.wait_for(
            connection.new_session(cwd=args.cwd, mcp_servers=[]), timeout=30
        )
        report["sessionId"] = session.session_id
        report["configOptions"] = [_dump(option) for option in session.config_options or []]
        prompt = await asyncio.wait_for(
            connection.prompt(
                session_id=session.session_id,
                prompt=[text_block("Reply with hello. Do not use tools.")],
            ),
            timeout=90,
        )
        report["prompt"] = _dump(prompt)
        report["ok"] = True
    except RequestError as exc:
        report["error"] = {
            "code": exc.code,
            "type": type(exc).__name__,
        }
    except Exception as exc:
        report["error"] = {"type": type(exc).__name__}
    finally:
        report["updates"] = updates
        if connection is not None:
            try:
                await asyncio.wait_for(connection.close(), timeout=3)
            except Exception:
                report["connection_close"] = False
        if process is not None:
            cleaned = terminate_process_group(process.pid, grace=3, trusted=True)
            try:
                await asyncio.wait_for(process.wait(), timeout=3)
            except asyncio.TimeoutError:
                cleaned = False
            report["cleanup"] = {
                "process_group_absent": bool(
                    cleaned and not process_group_alive(process.pid)
                )
            }
            if process.stdin is not None:
                process.stdin.close()
            if process.stderr is not None:
                stderr = await process.stderr.read()
                report["stderr_bytes"] = len(stderr)
        else:
            report["cleanup"] = {"process_group_absent": True}
    if not report["cleanup"]["process_group_absent"]:
        report["ok"] = False
    return report


def main(argv=None):
    parser = argparse.ArgumentParser()
    parser.add_argument("--command", required=True)
    parser.add_argument("--command-arg", action="append", default=[])
    parser.add_argument("--cwd", required=True)
    args = parser.parse_args(argv)
    report = asyncio.run(probe(args))
    print(json.dumps(report, ensure_ascii=False, sort_keys=True))
    return 0 if report["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/agents-orchestrator/tests/manual_acp_handshake.ts
#!/usr/bin/env bun

/** Probe an explicitly selected ACP Agent through the official TypeScript SDK. */

import { spawn, type ChildProcess } from "node:child_process";
import { Readable, Writable } from "node:stream";
import * as acp from "@agentclientprotocol/sdk";

import { createAcpClient } from "../scripts/backends/acp/client.ts";
import { processGroupAlive, terminateProcessGroup } from "../scripts/backends/acp/processes.ts";
import type { RuntimeRecord } from "../scripts/runtime_types.ts";

interface Options {
  command: string;
  commandArgs: string[];
  cwd: string;
}

function parseArguments(argv: string[]): Options {
  const options: Options = { command: "", commandArgs: [], cwd: "" };
  for (let index = 0; index < argv.length; index += 1) {
    const name = argv[index]!;
    const value = argv[index + 1];
    if (value === undefined) throw new Error(`${name} requires a value`);
    if (name === "--command") options.command = value;
    else if (name === "--command-arg") options.commandArgs.push(value);
    else if (name === "--cwd") options.cwd = value;
    else throw new Error(`unknown argument: ${name}`);
    index += 1;
  }
  if (!options.command || !options.cwd) throw new Error("--command and --cwd are required");
  return options;
}

function safeError(error: unknown): RuntimeRecord {
  const record: RuntimeRecord = { type: error instanceof Error ? error.name : "Error" };
  if (error !== null && typeof error === "object" && Number.isInteger((error as RuntimeRecord).code)) {
    record.code = Number((error as RuntimeRecord).code);
  }
  return record;
}

async function withTimeout<T>(operation: Promise<T>, milliseconds: number): Promise<T> {
  let timer: ReturnType<typeof setTimeout> | undefined;
  try {
    return await Promise.race([
      operation,
      new Promise<never>((_resolve, reject) => {
        timer = setTimeout(() => reject(new Error("TimeoutError")), milliseconds);
      }),
    ]);
  } finally {
    if (timer) clearTimeout(timer);
  }
}

export async function probe(options: Options): Promise<RuntimeRecord> {
  const updates: RuntimeRecord[] = [];
  const report: RuntimeRecord = { ok: false };
  let child: ChildProcess | null = null;
  let connection: acp.ClientConnection | null = null;
  try {
    child = spawn(options.command, options.commandArgs, {
      cwd: options.cwd,
      detached: true,
      stdio: ["pipe", "pipe", "pipe"],
    });
    if (!child.stdin || !child.stdout) throw new Error("ACP Agent did not expose stdio streams");
    const app = createAcpClient(
      () => ({ selectedOptionId: null, allowed: false }),
      (notification) => {
        updates.push({
          sessionId: notification.sessionId,
          type: notification.update.sessionUpdate,
        });
      },
    );
    connection = app.connect(acp.ndJsonStream(
      Writable.toWeb(child.stdin) as WritableStream<Uint8Array>,
      Readable.toWeb(child.stdout) as ReadableStream<Uint8Array>,
    ));
    const context = connection.agent;
    const initialized = await withTimeout(context.request(acp.methods.agent.initialize, {
      protocolVersion: acp.PROTOCOL_VERSION,
      clientCapabilities: {},
      clientInfo: { name: "agents-orchestrator-handshake", version: "1" },
    }), 30_000);
    report.initialized = {
      protocolVersion: initialized.protocolVersion,
      agentInfo: initialized.agentInfo ?? null,
      agentCapabilities: initialized.agentCapabilities ?? null,
      authMethods: initialized.authMethods ?? [],
    };
    const session = await withTimeout(context.request(acp.methods.agent.session.new, {
      cwd: options.cwd,
      mcpServers: [],
    }), 30_000);
    report.sessionId = session.sessionId;
    report.configOptions = session.configOptions ?? [];
    const prompted = await withTimeout(context.request(acp.methods.agent.session.prompt, {
      sessionId: session.sessionId,
      prompt: [{ type: "text", text: "Reply with hello. Do not use tools." }],
    }), 90_000);
    report.prompt = prompted;
    report.ok = true;
  } catch (error) {
    report.error = safeError(error);
  } finally {
    report.updates = updates;
    try { connection?.close(); } catch { report.connectionClose = false; }
    const pid = child?.pid;
    const cleaned = pid ? terminateProcessGroup(pid, { graceSeconds: 3, trusted: true }) : true;
    if (child?.stdin) child.stdin.destroy();
    if (child?.stderr) {
      const chunks: Buffer[] = [];
      child.stderr.on("data", (chunk: Buffer) => chunks.push(chunk));
      await Promise.race([
        new Promise<void>((resolve) => child!.stderr!.once("close", () => resolve())),
        Bun.sleep(500),
      ]);
      report.stderrBytes = Buffer.concat(chunks).byteLength;
    }
    report.cleanup = { processGroupAbsent: Boolean(cleaned && (!pid || !processGroupAlive(pid))) };
    if (!(report.cleanup as RuntimeRecord).processGroupAbsent) report.ok = false;
  }
  return report;
}

if (import.meta.main) {
  try {
    const report = await probe(parseArguments(process.argv.slice(2)));
    process.stdout.write(`${JSON.stringify(report)}\n`);
    process.exit(report.ok ? 0 : 1);
  } catch (error) {
    process.stderr.write(`manual_acp_handshake.ts: ${error instanceof Error ? error.message : "invalid arguments"}\n`);
    process.exit(2);
  }
}
